protocol/handshake.go

```diff
@@ -0,0 +1,122 @@
+package protocol
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+func NewHandshake(rw io.ReadWriter) *Handshake {
+	return &Handshake{rw: rw}
+}
+
+// Handshake handles the very first message passing of the protocol
+type Handshake struct {
+	rw     io.ReadWriter
+	passed bool
+}
+
+// Write implements MessageReadWriter
+func (h *Handshake) Write(m Message) error {
+	_, err := h.rw.Write(m)
+	return err
+}
+
+// Read implements MessageReadWriter
+func (h *Handshake) Read() (Message, error) {
+	if h.passed {
+		return h.readTypedMessage()
+	}
+	return h.readRawMessage()
+}
+
+// Init receives and validates the very first message from the frontend per session.
+// it may send message back to the frontend if needed to skip SSL request since it
+// currently not supported.
+//
+// once done, Init must not be called again, or error will be returned.
+func (h *Handshake) Init() (res Message, err error) {
+	if h.passed {
+		err = fmt.Errorf("handshake already passed")
+		return
+	}
+
+	// read the initial connection startup message
+	res, err = h.readRawMessage()
```

Comment:
to keep bit simpler, can also be:
```suggestion
	res, err = h.Read()
```
(same for line 63)